src/Shared/Traits.cs

```diff
@@ -13,7 +13,20 @@ namespace Microsoft.Build.Utilities
     /// </summary>
     internal class Traits
     {
-        public static Traits Instance = new Traits();
+        private static Traits _instance = new Traits();
+        public static Traits Instance
+        {
+            get
+            {
+#if DEBUG
+                if (Environment.GetEnvironmentVariable("MSBUILDRELOADTRAITSONEACHACCESS") == "1")
```

Comment:
I get why you did this, but I'm a little worried. Getting this environment variable will start eating up a lot of CPU on debug builds. We should try to think of another way to do this only for tests.

Maybe check executing assembly? BuildEnvironmentHelper already has a RunningTests property. Could do something similar to that.